File: 2024/4/4.py
f = open('input.txt').read()
lines = f.splitlines()

windows = [
    [
        "XMAS",
        "....",
        "....",
        "...."
    ],
    [
        "SAMX",
        "....",
        "....",
        "...."
    ],
    [
        "X...",
        "M...",
        "A...",
        "S..."
    ],
    [
        "S...",
        "A...",
        "M...",
        "X..."
    ],
    [
        "X...",
        ".M..",
        "..A.",
        "...S"
    ],
    [
        "S...",
        ".A..",
        "..M.",
        "...X"
    ],
    [
        "...S",
        "..A.",
        ".M..",
        "X..."
    ],
    [
        "...X",
        "..M.",
        ".A..",
        "S..."
    ]
]

windowsPart2 = [
    [
        "M.M",
        ".A.",
        "S.S"
    ],
    [
        "M.S",
        ".A.",
        "M.S"
    ],
    [
        "S.M",
        ".A.",
        "S.M"
    ],
    [
        "S.S",
        ".A.",
        "M.M"
    ]
]

def checkMatch(src, window, pos, maxHits):
    hits = 0
    for r in range(len(window)):
        if pos[0] + r >= len(src):
            continue
        for c in range(len(window[0])):
            if pos[1] + c >= len(src[0]):
                continue
            if window[r][c] == '.':
                continue
            if window[r][c] != src[pos[0] + r][pos[1] + c]:
                return False
            hits += 1
    return hits >= maxHits


def part1():
    matches = 0
    for row in range(len(lines)):
        for col in range(len(lines[0])):
            pos = (row, col)
            for window in windows:
                if checkMatch(lines, window, pos, 4):
                    matches += 1

    print(matches)

def part2():
    matches = 0
    for row in range(len(lines)):
        for col in range(len(lines[0])):
            pos = (row, col)
            for window in windowsPart2:
                if checkMatch(lines, window, pos, 5):
                    matches += 1

    print(matches)

part1()
part2()

File: 2024/4/4.hs

-- was unable to do it in haskell :(
-- wanted to do a sliding window solution but brain hurt due to recursion
-- so swapped to python for this week instead

main = do
    f <- readFile "input.txt"
    let ls = lines f
    print ls

window = ["XMAS", "....", "....", "...."]

-- match2 :: Eq a => [[a]] -> [[a]] -> (Int, Int) -> Maybe ()
-- match2 src window (row, col)
--     | length window > length src = Nothing
--     | otherwise = 

-- match :: src -> window -> (row, col)
match :: Eq a => [a] -> [a] -> Int-> Maybe ()
match src window n
    | length window > length src = Nothing
    | null window || null src = Just ()
    | head src == head window = match (tail src) (tail window) (n + 1)
    | otherwise = Nothing


safeAt2 :: [[a]] -> (Int, Int) -> Maybe a
safeAt2 src (row, col) = do
    r <- safeAt src row
    safeAt r col

safeAt :: [a] -> Int -> Maybe a
safeAt [] _ = Nothing
safeAt (x:xs) 0 = Just x
safeAt (x:xs) n
    | n < 0 = Nothing
    | otherwise = safeAt xs (n - 1)